README.md
# PhaseIIBeam_v68


History
Version 1 was the primary version with a few errors.

1/2/2018 Version 2: Changed the number of bins, minimum and maximum limits of histogram in proton.mac.

1/2/2018 Version 3: Changed the type of fNbElectronsStrip1, fNbElectronsStrip2 in RunAction.hh from G4double to G4int.

22/2/2018 Version 4: Added two couples of BPIX modules (one in front of and one behind the DUT). Calculating the entrance point of the primary proton inside Pixel Detector 1 and its exit point from Pixel Detector 2.

23/2/2018 Version 5: Decreased the thickness of pixel modules. Calculating the deflection angle of the primary track between the two pixel modules and filling with it histogram 84.

1/3/2018 Version 6: Local electric field added to each strip sensor. Global electric field could be added through ElectricFieldSetup.

7/3/2018 Version 7: Positions of primary inside strip sensors 1 and 2 displayed.

9/3/2018 Version 8: Midpoints of primary inside each detector displayed.

13/3/2018 Version 9: Let A = fPointPix1mid, B = fPointDet1mid, C = fPointDet2mid, D = fPointPix2mid. Let dB = distance between B and AD and dC = distance between C and AD. Histograms filled with dB and dC for each event.

13/3/2018 Version 10: Let B' be the AD point with same z as B. Let C' be the AD point with the same z as C. Histograms filled with B'Bx, B'By, C'Cx, C'Cy for each event.

13/3/2018 Version 11: Bug in filling the histograms per strip fixed.

15/3/2018 Version 12: World material changed to air.

20/3/2018 Version 13: Kinetic energy of secondaries calculated and could be printed out.

Versions 14 and 15 were non-working and full of bugs.

11/4/2018 Version 16: Possibility for 2S DUT rotation about the x-axis added.

11/4/2018 Version 17: Strips a and b for each sensor separated.

12/4/2018 Version 18: Changed "Air" material. Added CO2 and ArgonGas to it.

3/5/2018 Version 19: Changed titles of histograms 119-122.

Version 20 was the same as Version 19, but with different numbers of bins, min and max values in histograms.

25/5/2018 Version 21: Histograms and ntuples filled with number of hits per strip. We consider that when the deposited energy in the volume of a strip exceeds 5000 X 3.67 eV [5000 e = signal threshold for a valid hit, 3.67 eV = energy needed for the production of an electron-hole pair], then we have a hit.

21/6/2018 Version 22: Hit multiplicity redefined (per event). Scintillator added. Pixel layer added to pixel detectors.

25/6/2018 Version 22: Calculation of x- and y-coordinates of impact points.

Version 23 was a trial version to find the mistake in the rotation matrix of the DUT.

29/6/2018 Version 24: Calculation of z-coordinates of impact points.

8/7/2018 Version 25: Attempt to calculate stored energy inside each pixel-volume.

12/7/2018 Version 26: Calculation of number of hits per pixel module.

14/8/2018 Version 27: Calculation of average positions of primaries in Sensors 1 and 2 according to the centroids of the strip surfaces. Charge weighting included.

17/8/2018 Version 28: Check in every event if hits in Sensor 1 correspond to hits in Sensor2. Output printed after the end of each run. Bug with weights = 0 fixed. Pure detector resolution histograms added (reconstructed coordinate - real coordinate of primary particle in each sensor).

27/8/2018 Version 29: Impact points of primary protons in BPIX modules defined with respect to the centroids of pixels. The above positions in the pixel modules are plotted in histograms 203-208.

7/9/2018 Version 30: Seed selection option added in proton.mac.

8/9/2018 Version 31: Bug with fixing histograms 117 and 118 (which were filled with nan in a few cases) removed.

17/9/2018 Version 32: Weights of strips removed.

17/9/2018 Version 33: Bins of hit multiplicities fixed.

19/9/2018 Version 34: Reco-y splitted in three different histograms (y = -45 um, y = 0 um, y = 45 um).

27/9/2018 Version 35: Reco-y check per event option added: printing of every extrapolated and reconstructed position.

9/10/2018 Version 36: General particle source added. Changes made in PrymaryGeneratorAction and proton.mac. sigma of beam included.

10/10/2018 Version 37: Changed beam shape to ellipsoid and removed sigma.

21/12/2019 Version 49: This version is based on v37, ignoring the previous versions which were using geometries for the telescope planes different than those of the Phase-1 BPIX modules. Active thickness for strip sensors switched to 290 um, as these sensors will be used after all. Modules doubled in y-direction (although no pixel information available close to the edges: it's easier to switch between one- and two-module planes if avoiding to send the beam close to the edges of the plane). gps position = (-1, -0.1, -20) cm

22/12/2019 Version 50: Hit match finding added in 2S for a window of 5 strips. No check for distance to track yet.

23/12/2019 Version 51: Aluminum added in front of particle beam. Number of events with hit matches in 2S sensors printed out.

25/12/2019 Version 52: Positions of modules, scintillators, DUT and aluminium changed. z = 0 cm corresponds to z = 13.5 cm in real life. Second scintillator added.

26/12/2019 Version 53: Calculation of energy deposition in scintillators added.

28/12/2019 Version 53: Stub defined: If hit matching in 1st 2S sensor < 100 um away from track in y-direction, then we have efficiency = 1 for the current event.

3/1/2020 Version 54: xy coordinates of beam 20 cm after the last scintillator plotted.

4/1/2020 Version 55: Test version.

6/1/2020 Version 56: FWHM = 5 mm, halfx = halfy = 2.123 mm. 

9/1/2020 Version 59: Rotation of BPIX removed. (It was 0 anyway.)

9/1/2020 Version 60: World Size changed to 47 cm.

10/1/2020 Version 61: Calculation of energy deposition inside each component added.

11/1/2020 Version 62: Calculation of kinetic energy at entrance and exit of each component added.

11/1/2020 Version 63: Test and continuation of version 62.


src/StackingAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file electromagnetic/TestEm18/src/StackingAction.cc
/// \brief Implementation of the StackingAction class
//
// $Id: StackingAction.cc 67268 2013-02-13 11:38:40Z ihrivnac $
//
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "StackingAction.hh"

#include "DetectorConstruction.hh"
#include "EventAction.hh"
#include "HistoManager.hh"
#include "RunAction.hh"

#include "G4Step.hh"
#include "G4Track.hh"
#include "G4VTrajectory.hh"
#include "G4TransportationManager.hh"
#include "G4Navigator.hh"
#include "G4PVPlacement.hh"
#include "G4GeometryManager.hh"
#include "G4PhysicalVolumeStore.hh"
#include "G4LogicalVolumeStore.hh"
#include "G4SolidStore.hh"

#include "G4Positron.hh"
#include "G4Electron.hh"
#include "G4PionPlus.hh"
#include "G4PionMinus.hh"
#include "G4MuonPlus.hh"
#include "G4MuonMinus.hh"
#include "G4Gamma.hh"

#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"
#include "G4VisAttributes.hh"

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

StackingAction::StackingAction(DetectorConstruction* DA, RunAction* RA, EventAction* EA)
:G4UserStackingAction(), fDetectorconstruction(DA), fRunaction(RA), fEventaction(EA)
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

StackingAction::~StackingAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

G4ClassificationOfNewTrack
StackingAction::ClassifyNewTrack(const G4Track* track)
{
  //define speed unit
  new G4UnitDefinition ("meter/second", "m/s", "Speed", m/s);
  
  //energy spectrum, polarization, initial momentum direction, track weight, velocity and local time of secondaries and tertiaries
  //
  energy = track->GetKineticEnergy();
  xpolarization = track->GetPolarization().x();
  ypolarization = track->GetPolarization().y();
  zpolarization = track->GetPolarization().z();
  charged = (track->GetDefinition()->GetPDGCharge() != 0.);
  G4VPhysicalVolume* volume = track->GetVolume();
  G4double trackWeight = track->GetWeight();
  G4double initVelocity = track->GetVelocity();
  G4double kinEnergy = track->GetKineticEnergy();
  G4double time = track->GetLocalTime();

  //geometry parameters
  G4double Det1SizeZ, Det2SizeZ, Dist, Strip1Depth, Strip1Length, Strip2Depth, Strip2Length, StripDist, StripWidth, xAbs, XangleDUT, XangleBPIX, YangleBPIX, theta;
  G4double posZAl, posZScint1, posZScint2, posZBPIX12, posZBPIX34, posZ1, posZ2, posZDUT;
  G4int div;
  G4int totalNbStrips, stripNo;
  Det1SizeZ=fDetectorconstruction->GetSize1();
  Det2SizeZ=fDetectorconstruction->GetSize2();
  Dist=fDetectorconstruction->GetDist();
  Strip1Depth=fDetectorconstruction->GetStrip1Depth();
  Strip1Length=fDetectorconstruction->GetStrip1Length();
  Strip2Depth=fDetectorconstruction->GetStrip2Depth();
  Strip2Length=fDetectorconstruction->GetStrip2Length();
  StripDist=fDetectorconstruction->GetStripDist();
  StripWidth=fDetectorconstruction->GetStripWidth();
  posZAl=fDetectorconstruction->GetPosZAl();
  posZScint1=fDetectorconstruction->GetPosZScint1();
  posZScint2=fDetectorconstruction->GetPosZScint2();
  posZBPIX12=fDetectorconstruction->GetPosZBPIX12();
  posZBPIX34=fDetectorconstruction->GetPosZBPIX34();
  posZ1=fDetectorconstruction->GetPosZ1();
  posZ2=fDetectorconstruction->GetPosZ2();
  posZDUT=fDetectorconstruction->GetPosZDUT();
  XangleDUT=fDetectorconstruction->GetDUTangleX();
  XangleBPIX=fDetectorconstruction->GetBPIXangleX();
  YangleBPIX=fDetectorconstruction->GetBPIXangleY();
  totalNbStrips = 1016;

  //G4ThreeVector primVert = track->GetVertexPosition();
  G4double x_prim, y_prim, z_prim, xPix1, yPix1, zPix1, xPix, yPix, zPix, x_after, y_after, z_after;
  x_after = track->GetVertexPosition().x();
  y_after = track->GetVertexPosition().y();
  z_after = track->GetVertexPosition().z();
  //position if the DUT wasn't rotated
  x_prim = x_after;
  y_prim = y_after*cos(XangleDUT) - z_after*sin(XangleDUT);
  z_prim = y_after*sin(XangleDUT) + z_after*cos(XangleDUT);
  //position if the BPIX modules weren't rotated
  xPix1 = x_after;
  yPix1 = y_after;
  zPix1 = z_after;
  xPix = x_after;
  yPix = y_after;
  zPix = z_after;


  //secret number
  G4int iSecret, jSecret;

  //G4cout 
      //<< "\n Local time: " 
      //<< G4BestUnit(time, "Time")
      //<< "\n Initial velocity: " 
      //<< G4BestUnit(initVelocity, "Speed")
      //<< "\n Kinetic energy: "
      //<< G4BestUnit(kinEnergy, "Energy")
      //<< "\n Track weight: " 
      //<< trackWeight
      //<< G4endl;

  //particle definition
  const G4ParticleDefinition* particleDefinition = track->GetParticleDefinition();

  G4ThreeVector initMom = track->GetMomentum();

  //polarization of secondaries and tertiaries calculation
  if (track->GetParentID() == 1)  {
    fEventaction->AddSecondaryxPolarization(xpolarization);
    fEventaction->AddSecondaryyPolarization(ypolarization ); 
    fEventaction->AddSecondaryzPolarization(zpolarization); 
  }

  if (track->GetParentID() > 1)  {
    fEventaction->AddTertiaryxPolarization(xpolarization); 
    fEventaction->AddTertiaryyPolarization(ypolarization); 
    fEventaction->AddTertiaryzPolarization(zpolarization); 
  }

  //kinetic energy of secondaries calculation 
  if (track->GetParentID() == 1)  {

    if (volume == fDetectorconstruction->GetWorld()) {

      if (z_prim < posZDUT - Dist/2 - Det1SizeZ)  {

      	if(particleDefinition == G4Positron::Definition())   {
          fRunaction->AddNbPositronsB1();
      	}

      	if(particleDefinition == G4Electron::Definition())   {
          fRunaction->AddNbElectronsB1();
      	}

      }
      if ((z_prim > posZDUT - Dist/2) && (z_prim < posZDUT + Dist/2))  {

      	if(particleDefinition == G4Positron::Definition())   {
          fRunaction->AddNbPositrons12();
      	}

      	if(particleDefinition == G4Electron::Definition())   {
          fRunaction->AddNbElectrons12();
      	}

      }
      if (z_prim > posZDUT + Dist/2 + Det2SizeZ)  {

      	if(particleDefinition == G4Positron::Definition())   {
          fRunaction->AddNbPositronsA2();
      	}

      	if(particleDefinition == G4Electron::Definition())   {
          fRunaction->AddNbElectronsA2();
      	}

      }

      //G4cout 
          //<< "\n Secondary produced with initial momentum direction:" 
          //<< initMom
          //<< G4endl;
    }

    if (volume == fDetectorconstruction->GetDet1()) {
      G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
      /*G4cout
         << "\n particle from 2S, outside sensors"
         << G4endl;*/
  
      if (charged) {
       fRunaction->AddChargedSecondary1(energy);
       analysisManager->FillH1(9,energy);
      } else {
       fRunaction->AddNeutralSecondary1(energy);
       analysisManager->FillH1(11,energy);
      }

      fEventaction->AddSecondary1(energy); 

      if(particleDefinition == G4Positron::Definition())   {
        fRunaction->AddNbPositrons1();
      }

      if(particleDefinition == G4Electron::Definition())   {
        fRunaction->AddNbElectrons1();
        fEventaction->AddNb2SElectrons();
        G4cout
            << "\n electron from 2S"
            << G4endl;
	if ((z_prim >= posZDUT - Dist/2 - Det1SizeZ) && (z_prim <= posZDUT - Dist/2 - Det1SizeZ + Strip1Depth) && (x_prim >= (-Strip1Length/2)) && (x_prim <= Strip1Length/2)) {
            /*G4cout
               << "\n electron from 2S"
               << G4endl; */
            div = y_prim/(StripWidth+StripDist);
	    if (y_prim == 0)   {
	       iSecret = rand() % 99;
	       if (iSecret < 50)   {
		  stripNo = totalNbStrips/2 + div;
	       }
	       if (iSecret >= 50)   {
	       	  stripNo = totalNbStrips/2 + 1 + div;
	       }
	    }	       
            if (y_prim > 0)    {
	       stripNo = totalNbStrips/2 + 1 + div;
            }
            if (y_prim < 0)    {
               stripNo = totalNbStrips/2 + div;
            }
            if ((stripNo <= totalNbStrips) && (stripNo > 0))  {
     	    //G4cout
	       //<< "\n Secondary electron created below strip No "
               //<< stripNo
               //<< "\n y = "
	       //<< y_prim
               //<< "\n div = "
	       //<< div
               //<< G4endl;
               if (x_prim > 0)  {
	          fRunaction->AddNbElectronsStrip1a(stripNo);
               }
               if (x_prim < 0)  {
	          fRunaction->AddNbElectronsStrip1b(stripNo);
               }
               if (x_prim == 0)  {
	       	jSecret = rand() % 99;
	       	if (jSecret < 50)   {
	            fRunaction->AddNbElectronsStrip1a(stripNo);
	       	}
	       	if (jSecret >= 50)   {
	            fRunaction->AddNbElectronsStrip1b(stripNo);
	       	}
               }
            }
        }
      }

      if(particleDefinition == G4PionPlus::Definition())   {
        fRunaction->AddNbPionsPlus1();
      }

      if(particleDefinition == G4PionMinus::Definition())   {
        fRunaction->AddNbPionsMinus1();
      }

      if(particleDefinition == G4MuonPlus::Definition())   {
        fRunaction->AddNbMuonsPlus1();
      }

      if(particleDefinition == G4MuonMinus::Definition())   {
        fRunaction->AddNbMuonsMinus1();
      }

      //G4cout 
          //<< "\n Secondary produced with initial momentum direction:" 
          //<< initMom
          //<< G4endl;
    }

    if (volume == fDetectorconstruction->GetDet2()) {
      G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
      /*G4cout
         << "\n particle from 2S, outside sensors"
         << G4endl;*/
  
      if (charged) {
       fRunaction->AddChargedSecondary2(energy);
       analysisManager->FillH1(10,energy);
      } else {
       fRunaction->AddNeutralSecondary2(energy);
       analysisManager->FillH1(12,energy);
      }

      fEventaction->AddSecondary2(energy); 

      if(particleDefinition == G4Positron::Definition())   {
        fRunaction->AddNbPositrons2();
      }

      if(particleDefinition == G4Electron::Definition())   {
        fRunaction->AddNbElectrons2();
        fEventaction->AddNb2SElectrons();
        G4cout
            << "\n electron from 2S"
            << G4endl;
	if ((z_prim >= posZDUT + Dist/2) && (z_prim <= (posZDUT + Dist/2 + Strip2Depth))&& (x_prim >= (-Strip2Length/2)) && (x_prim <= Strip2Length/2)) {
            /*G4cout
               << "\n electron from 2S"
               << G4endl;*/
            div = y_prim/(StripWidth+StripDist);
	    if (y_prim == 0)   {
	       iSecret = rand() % 99;
	       if (iSecret < 50)   {
		  stripNo = totalNbStrips/2 + div;
	       }
	       if (iSecret >= 50)   {
	       	  stripNo = totalNbStrips/2 + 1 + div;
	       }
	    }
            if (y_prim > 0)    {
	       stripNo = totalNbStrips/2 + 1 + div;
            }
            if (y_prim < 0)    {
               stripNo = totalNbStrips/2 + div;
            }
            if ((stripNo <= totalNbStrips) && (stripNo > 0))  {
     	    //G4cout
	       //<< "\n Secondary electron created below strip No "
               //<< stripNo
               //<< "\n y = "
	       //<< y_prim
               //<< "\n div = "
	       //<< div
               //<< G4endl;
               if (x_prim > 0)  {
	          fRunaction->AddNbElectronsStrip2a(stripNo);
               }
               if (x_prim < 0)  {
	          fRunaction->AddNbElectronsStrip2b(stripNo);
               }
               if (x_prim == 0)  {
	       	jSecret = rand() % 99;
	       	if (jSecret < 50)   {
	            fRunaction->AddNbElectronsStrip2a(stripNo);
	       	}
	       	if (jSecret >= 50)   {
	            fRunaction->AddNbElectronsStrip2b(stripNo);
	       	}
               }
            }
        }
      }

      if(particleDefinition == G4PionPlus::Definition())   {
        fRunaction->AddNbPionsPlus2();
      }

      if(particleDefinition == G4PionMinus::Definition())   {
        fRunaction->AddNbPionsMinus2();
      }

      if(particleDefinition == G4MuonPlus::Definition())   {
        fRunaction->AddNbMuonsPlus2();
      }

      if(particleDefinition == G4MuonMinus::Definition())   {
        fRunaction->AddNbMuonsMinus2();
      }

      //G4cout 
          //<< "\n Secondary produced with initial momentum direction:" 
          //<< initMom
          //<< G4endl;
    }

  }

  //kinetic energy of tertiaries calculation 
  if (track->GetParentID() > 1)  {
    if (volume == fDetectorconstruction->GetDet1()) {
      G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  
      if (charged) {
       fRunaction->AddChargedTertiary1(energy);
       analysisManager->FillH1(13,energy);
      } else {
       fRunaction->AddNeutralTertiary1(energy);
       analysisManager->FillH1(15,energy);
      }

      fEventaction->AddTertiary1(energy);

      if(particleDefinition == G4Positron::Definition())   {
        fRunaction->AddNbPositrons1();
      }

      if(particleDefinition == G4Electron::Definition())   {
        fRunaction->AddNbElectrons1();
      }

      if(particleDefinition == G4PionPlus::Definition())   {
        fRunaction->AddNbPionsPlus1();
      }

      if(particleDefinition == G4PionMinus::Definition())   {
        fRunaction->AddNbPionsMinus1();
      }

      if(particleDefinition == G4MuonPlus::Definition())   {
        fRunaction->AddNbMuonsPlus1();
      }

      if(particleDefinition == G4MuonMinus::Definition())   {
        fRunaction->AddNbMuonsMinus1();
      } 
    }

    if (volume == fDetectorconstruction->GetDet2()) {
      G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  
      if (charged) {
       fRunaction->AddChargedTertiary2(energy);
       analysisManager->FillH1(14,energy);
      } else {
       fRunaction->AddNeutralTertiary2(energy);
       analysisManager->FillH1(16,energy);
      }

      fEventaction->AddTertiary2(energy); 

      if(particleDefinition == G4Positron::Definition())   {
        fRunaction->AddNbPositrons2();
      }

      if(particleDefinition == G4Electron::Definition())   {
        fRunaction->AddNbElectrons2();
      }

      if(particleDefinition == G4PionPlus::Definition())   {
        fRunaction->AddNbPionsPlus2();
      }

      if(particleDefinition == G4PionMinus::Definition())   {
        fRunaction->AddNbPionsMinus2();
      }

      if(particleDefinition == G4MuonPlus::Definition())   {
        fRunaction->AddNbMuonsPlus2();
      }

      if(particleDefinition == G4MuonMinus::Definition())   {
        fRunaction->AddNbMuonsMinus2();
      }
    }
  }

  return fUrgent;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
